### ISSUE TITLE
Implement natural sorting

### ISSUE THREAD

**Issue by garster** [LycheeOrg/Lychee-Laravel#321]
Album and photo title sort uses the database order by clause. This means a title that contain a sequence of numbers will not sort naturally. For example 1, 2, 10 will comes out at 1, 10, 2 due to how MySql/MariaDB orders string values.

It would be super cool if the [sorting was natural](https://en.wikipedia.org/wiki/Natural_sort_order). There are some natural sorting tools out there like PHP's natsort or Laravel sortBy with a callback. Or it could happen in the JS frontend. 

It would add some overhead so maybe it would be an option to turn on, or only enabled on string values. 

I could see this being of value for people that use the albums for things like addresses (real estate) or numbered landmarks, really anything where a number is used in the name.

![image](https://user-images.githubusercontent.com/4443025/63218927-8ae43480-c11b-11e9-9b12-08efc3365b47.png)

![image](https://user-images.githubusercontent.com/4443025/63218928-93d50600-c11b-11e9-853e-aa13d22e1cf1.png)

**Comment by garster** [comment]
I found a way to implement it for subalbums. I will need to clean up, do it for top level albums and maybe I can do a PR.

Before
![image](https://user-images.githubusercontent.com/4443025/63220069-b9bad480-c134-11e9-81d6-4d6c4c58be9e.png)





After
![image](https://user-images.githubusercontent.com/4443025/63220061-97c15200-c134-11e9-89b4-49ee5bf897ec.png)

**Comment by kamil4** [comment]
Thanks for taking the initiative on this! While I can see the potential value of this feature, to me personally it is not a priority, though other developers might feel different.

I wonder if we should replace the sorting we have with the natural sort or offer it as an option? Making it optional is of course safer; however, extending the GUI can be an order-of-magnitude more work than just working on the feature on the server in the first place... @ildyria, @d7415, what do you think? My feeling is that we should see performance of the two sorts on an album with >1,000 pictures and decide then.. You'll also first need to verify that Laravel's natural sort works correctly  on all datatypes of interest (some of the columns we can sort by contain booleans, numbers, or even dates).

Implementation-wise, you know already where to insert the code to sort subalbums. For top-level albums, I would think that modifying the `getToplevelAlbums` method in `ModelFunctions/AlbumFunctions.php` should be enough. And `get_albums_user` in the same file, I guess... However, to make the implementation complete, you should also add support for natural sorting of photos, not just albums. That will probably be `set_order` in `Photo.php` as well as `gen_thumbs` in `ModelFunctions/AlbumFunctions.php`.

**Comment by d7415** [comment]
I agree with @kamil4 re performance. If it's comparable then I'd default to it, if it's slower than an option may be required.

I wonder whether [this](https://stackoverflow.com/a/8557283/1777985) dirty hack might be of any use.

From the link:
```sql
order by lpad(name, 10, 0)  # assuming maximum string length is 10
```

**Comment by garster** [comment]
I created a very crude speed test. I created a mysql table with varchar(36). I then generated and inserted 10,000 UUIDs.

SELECT took 0.006s
SELECT with ORDER BY took 0.016s
PHP Sort took 0.001s
PHP Sort with Natural flag took 0.001s

I ran this on Windows using the PHP CLI using the MySQLi driver. I am not set up for PHP dev so that is why this is crude.

The Sort timing was only on the sort, but I did do a fresh select each time. I ran each test several times and that is the average I saw. Not super scientific, but it was apples to apples. To my surprise the sort was way faster. My test might be flawed, but some other validation would be good. I can share my embarrassing code.

**Comment by kamil4** [comment]
Thank you, this certainly looks encouraging! While natural sort appears to be an order of magnitude slower than the regular PHP sort, they are both pretty much in the noise compared to the time SQL queries take, even more so if `ORDER BY` is used.

**Comment by kamil4** [comment]
> I wonder whether [this](https://stackoverflow.com/a/8557283/1777985) dirty hack might be of any use.
> 
> From the link:
> 
> ```sql
> order by lpad(name, 10, 0)  # assuming maximum string length is 10
> ```

I think that would be appropriate only if we knew something about the structure of the alphanumeric field, which in case of Lychee we don't since it's supplied by the user...

**Comment by garster** [comment]
I realized in my testing that there is no index on that column so that could explain the order by slowness. Yes in the Lychee DB the only indexes are on id and parentId so it's still a somewhat valid comparison.

**Comment by kamil4** [comment]
Closed via #322. Thanks!
